Extension/src/LanguageServer/settings.ts

```diff
@@ -996,28 +902,76 @@ export class OtherSettings {
+                return minimum;
+            }
+            if (maximum !== undefined && value > maximum) {
+                return maximum;
+            }
+            return value;
+        }
+        const setting = section.inspect<number>(settingName);
+        return setting?.defaultValue ?? defaultValue;
+    }
 
-    public get customTextMateRules(): TextMateRule[] | undefined { return vscode.workspace.getConfiguration("editor.tokenColorCustomizations").get<TextMateRule[]>("textMateRules"); }
-    public getCustomThemeSpecificTextMateRules(themeName: string): TextMateRule[] | undefined { return vscode.workspace.getConfiguration(`editor.tokenColorCustomizations.[${themeName}]`, this.resource).get<TextMateRule[]>("textMateRules"); }
+    private getAsAssociations(sectionName: string, settingName: string, resource?: any): Associations {
+        const section = vscode.workspace.getConfiguration(sectionName, resource);
+        const value = section.get<any>(settingName);
+        if (isValidMapping(value, 'string', 'string')) {
+            return value as Associations;
+        }
+        const setting = section.inspect<any>(settingName);
+        return setting?.defaultValue as Associations;
+    }
+
+    private getAsExcludes(sectionName: string, settingName: string, resource?: any): Excludes {
+        const section = vscode.workspace.getConfiguration(sectionName, resource);
+        const value = section.get<any>(settingName);
+        if (isValidMapping(value, 'string', 'boolean')) {
+            return value as Excludes;
+        }
+        const setting = section.inspect<any>(settingName);
+        return setting?.defaultValue as Excludes;
+    }
+
+    // All default values are obtained from the VS Code settings UI. Please update the default values as needed.
+    public get editorTabSize(): number { return this.getAsNumber("editor", "tabSize", this.resource, 4, 1); }
+    public get editorInsertSpaces(): boolean { return this.getAsBoolean("editor", "insertSpaces", this.resource, true); }
+    public get editorAutoClosingBrackets(): string { return this.getAsString("editor", "autoClosingBrackets", this.resource, "languageDefined"); }
+    public get filesEncoding(): string { return this.getAsString("files", "encoding", { uri: this.resource, languageId: "cpp" }, "utf8"); }
+    public get filesAssociations(): Associations { return this.getAsAssociations("files", "associations"); }
+    public set filesAssociations(value: any) { void vscode.workspace.getConfiguration("files").update("associations", value, vscode.ConfigurationTarget.Workspace); }
+    public get filesExclude(): Excludes { return this.getAsExcludes("files", "exclude", this.resource); }
+    public get filesAutoSaveAfterDelay(): boolean { return this.getAsString("files", "autoSave", this.resource, "off") === "afterDelay"; }
+    public get editorInlayHintsEnabled(): boolean { return this.getAsString("editor.inlayHints", "enabled", this.resource, "on") !== "off"; }
```

Comment:
Do we need to be providing default value here?  (i.e. "on").  Doesn't VS Code provide a value in `defaultValue` from the call to `inspect`?  VS Code appears to know the default is `on` in the settings UI (and that `ui` is the default for `workbench.settings.editor`, etc.) If that is not being returned correctly by `inspect`, could you open a bug against VS Code to track that?

Reply:
This is just for the worst-case scenario where somehow there isn't a default value found. I don't think this ever happens but since we don't control the VS Code settings we want to provide our own defaults for security.

Reply:
Its okay for our own settings to sometime not have hardcoded defaults because we are more aware of the behavior we expect.

Reply:
I'm not sure I can buy into that reasoning.  If the existing functionality in VS Code were to break, that would be a bug, and would affect more than just us.  If this is a failsafe, could you please add code here that at least logs (if not sends telemetry for) when then failsafe is necessary, so we could report our findings as bugs against VS Code?

Reply:
Hard-coding someone else's defaults doesn't seem particular safe or robust.  If we keep this, the logging I suggested should also confirm that our hard-coded defaults (still) match the defaults provided by VS Code.  Seems like bugs could arise if trying to apply defaults that aren't correct and don't match the behavior of VS Code.

Reply:
I can add logging that tells the user we are using a default that is different than we expect. This would check if the VS Code setting default is undefined, tell the user which setting is returning the unexpected default and let them know that we are currently falling back to our own expected default. 

I understand what you are saying. The question is do we prefer to depend on VS Code here or make this our responsibility to make sure the defaults are defined correctly. 

@bobbrow What do you think?

Reply:
Actually, I believe that hard coding a default is more robust than not. If the setting were to be renamed and basically "disappear", our extension would continue to have defined behavior until we have a chance to update it to react to the change.  If the setting's type or enum values change, we can fall back to a default that we know we can handle.

However, I do think that logging telemetry to notify us of such a change would be very wise and give us advance notification of any such changes.